File: ios/Runner/MediaPipe/UprightRgbaBridge.swift
import CoreGraphics
import Flutter
import Foundation
import UIKit

/// Decodes a `FrameArgs` payload into upright RGBA8888 bytes plus the upright
/// dimensions. Used by the Dart side to feed analyzers that expect rotation=0
/// RGBA (ML Kit eye contour + EyeOcclusionUtil pixel analysis).
enum UprightRgbaBridge {
    static func decode(frame: FrameArgs) -> [String: Any]? {
        guard let image = FrameDecoder.decodeUprightImage(frame: frame),
              let cg = image.cgImage else {
            return nil
        }
        let width = cg.width
        let height = cg.height
        let bytesPerRow = width * 4
        var pixels = [UInt8](repeating: 0, count: width * height * 4)
        let colorSpace = CGColorSpaceCreateDeviceRGB()
        let bitmapInfo = CGBitmapInfo(rawValue: CGImageAlphaInfo.premultipliedLast.rawValue)
        guard let ctx = pixels.withUnsafeMutableBytes({ raw -> CGContext? in
            CGContext(
                data: raw.baseAddress,
                width: width,
                height: height,
                bitsPerComponent: 8,
                bytesPerRow: bytesPerRow,
                space: colorSpace,
                bitmapInfo: bitmapInfo.rawValue,
            )
        }) else {
            return nil
        }
        ctx.draw(cg, in: CGRect(x: 0, y: 0, width: width, height: height))
        return [
            "bytes": FlutterStandardTypedData(bytes: Data(pixels)),
            "width": width,
            "height": height,
        ]
    }
}

File: android/app/src/main/kotlin/com/example/face_liveness_demo/mediapipe/UprightRgbaBridge.kt
package com.example.face_liveness_demo.mediapipe

import java.nio.ByteBuffer

/**
 * Decodes a [FrameArgs] payload into upright RGBA8888 bytes plus the upright
 * dimensions. Reused by the Dart side to feed ML Kit / pixel-analysis paths
 * that expect rotation=0 RGBA (the easy path on both ML Kit and our
 * EyeOcclusionUtil pixel logic).
 *
 * Returns null when decode fails (unsupported format / corrupt bytes).
 */
object UprightRgbaBridge {
    data class Result(val bytes: ByteArray, val width: Int, val height: Int)

    fun decode(frame: FrameArgs): Result? {
        val bitmap = FrameDecoder.decodeUprightBitmap(frame) ?: return null
        try {
            val w = bitmap.width
            val h = bitmap.height
            val argb = IntArray(w * h)
            bitmap.getPixels(argb, 0, w, 0, 0, w, h)
            val rgba = ByteArray(w * h * 4)
            var dst = 0
            for (px in argb) {
                rgba[dst] = ((px shr 16) and 0xFF).toByte()      // R
                rgba[dst + 1] = ((px shr 8) and 0xFF).toByte()   // G
                rgba[dst + 2] = (px and 0xFF).toByte()           // B
                rgba[dst + 3] = ((px shr 24) and 0xFF).toByte()  // A
                dst += 4
            }
            return Result(rgba, w, h)
        } finally {
            bitmap.recycle()
        }
    }

    /** Builds a Flutter-compatible map result. */
    fun toMap(result: Result): Map<String, Any> = mapOf(
        "bytes" to result.bytes,
        "width" to result.width,
        "height" to result.height,
    )

    @Suppress("unused")
    private fun discardBuffer(buf: ByteBuffer) {
        // Reserved for future direct-buffer optimisation. No-op for now.
    }
}
